BasicRPG/BasicRPG/Program.cs

```diff
@@ -1,160 +1,121 @@
-            player = new Mage();
+            case 1: 
+                return new Warrior(name);
+            case 2:
+                return new Mage(name);   
+            default:
+                return new Idiot(name);                
         }
-        else if (characterChoice == 2)
-        {
-            Console.WriteLine($"From here on out you are {name} the cunning mage!");
-            player = new Warrior();
-        }
-
-        return player;
     }
-    static void battle(Character player, Character enemy)
+    static void Battle(Character player, Character enemy) 
     {
         Random random = new Random();
-        bool playerAlive = true;
-        bool enemyAlive = true;
-        int playerDamage;
-        int enemyDamage;
-        Console.WriteLine("You are now fighting the mighty GOBLIN!");
+        Console.WriteLine($"{player.name} is fighting {enemy.name}!");
 
-        while(playerAlive && enemyAlive)
+        while(player.IsAlive() && enemy.IsAlive()) 
         {
-            playerDamage = random.Next(1, 6);
-            enemyDamage = random.Next(1, 6);
             Console.WriteLine("Rolling the dice for player damage!");
             Console.ReadKey();
-            player.dealDamage(enemy, playerDamage);
+            player.DealDamage(enemy, random.Next(1, 6));
+            if (!enemy.IsAlive()) 
+            {
+                Console.WriteLine($"{enemy.name} has DIED! You are victorious!");
+                return;
+            }
             Console.WriteLine("Rolling the dice for enemy damage!");
             Console.ReadKey();
-            enemy.dealDamage(player, enemyDamage);
+            enemy.DealDamage(player, random.Next(1, 6));
+            if (!player.IsAlive()) 
+            {
+                Console.WriteLine($"{player.name} has DIED! You lost!");
+                return;
+            }
             Console.ReadKey();
-            playerAlive = player.isAlive();
-            enemyAlive = enemy.isAlive();
-        }
-
-        if (playerAlive == false)
-        {
-            Console.WriteLine("The player has DIED! You lost!");
-        }
-        else if (enemyAlive == false)
-        {
-            Console.WriteLine("You defeated the enemy! You are victorious!");
         }
-
-        Console.ReadKey();
-  
     }
-
 }
 
-public class Character
+public class Character 
 {
-    public bool alive;
-    public int playerHp = 20;
-    public int enemyHp = 20;
-    string name = " ";
-    public virtual void takeDamage(Character x)
-    {
+    int hp = 20;
+    public string name = "";
 
+    public Character(string name)
+    {
+        this.name = name;
     }
-    public virtual void dealDamage(Character x, int dmg)
+    public void DealDamage(Character defender, int damage) 
     {
+        defender.setHp(defender.hp - damage);
+        Console.WriteLine($"{name} has dealt {damage} damage to {defender.name}! {defender.name} has {defender.hp} hp left!");
+    }
```

Comment:
You don't need both `dealDamage` and `takeDamage` on the character class, since they both do the same thing. Instead, make `dealDamage` a method of the `Character` class, which takes two params: the character that is taking the damage (`defender`) and the amount of `damage` the character is dealing.

Think of `dealDamage` as something that the `Character` class is able to do. Rather, methods are things your class is able to do. So now you can call it like this:

```cs
player.dealDamage(enemy, 5);
enemy.dealDamage(player, 5);
```

Having one method to deal the damage allows for more funky stuff to happen, such as a `Character` instance doing damage to themselves.

```cs
player.dealDamage(player, 5);
enemy.dealDamage(enemey, 5);
```

There are multiple logic paths you can think of to implement this, this is just one of them.

Reply:
Here, `name` refers to the name of the current character class that this method is executing on. For example, for this call of the method:

```cs
player.dealDamage(enemy, 5);
```

`name` is `player.name`. and `defender` is `enemy`

Because we are calling `name` from within the `Character` class for the `player` instance, we don't specify that we're using `player.name`, since the class itself doesn't know that. The instance however does.